examples/bond/assets/bond-dapp-stateful.py

```diff
@@ -0,0 +1,206 @@
+                Gtxn[1].sender() == App.globalGet(Bytes("buyback")),
+                # verify amount of algo received by buyer
+                Gtxn[1].amount() == Minus(Mul(Gtxn[0].asset_amount(), App.globalGet(nominal_price)), Gtxn[1].fee()),
+                Gtxn[1].receiver() == Txn.sender(),
+                # verify maturity date is passed
+                Global.latest_timestamp() > App.globalGet(maturity_date)
+            )
+        ),
+        Return(Int(1))
+    ])
+
+    on_redeem_coupon = Seq([
+        Assert(
+            And(
+                basic_checks,
+                # User sends `B_i` to `DEX_i` lsig.
+                Gtxn[0].type_enum() == TxnType.AssetTransfer,
+                # `DEX_i` sends `B_{i+1}` to the user.
+                Gtxn[1].type_enum() == TxnType.AssetTransfer,
+                Gtxn[0].asset_amount() == Gtxn[1].asset_amount(),
+                # `Dex_i` sends coupon value to user
+                Gtxn[2].type_enum() == TxnType.Payment,
+                Gtxn[2].receiver() == Txn.sender(),
+                # verify coupon amount
+                Gtxn[2].amount() == Mul(Gtxn[0].asset_amount(), App.globalGet(coupon_value))
+            )
+        ),
+        Return(Int(1))
+    ])
+
+    # fetch asset_holding.balance from Txn.accounts[0]
+    asset_balance = AssetHolding.balance(Int(1), Gtxn[1].xfer_asset())
+
+    on_create_dex = Seq([
+        asset_balance, # load asset_balance from store
+        Assert(
+            And(
+                Txn.sender() == App.globalGet(store_manager),
+                basic_checks,
+                Gtxn[0].type_enum() == TxnType.AssetTransfer,
+                # transfer `balanceOf(issuer, B_i)`  of `B_{i+1}` from the creator to the `issuer`.
+                # index 1 of Txn.accounts().
+                asset_balance.value() == Gtxn[0].asset_amount(),
+                # burn `B_i` issuer bonds. send to null_address
+                Gtxn[1].type_enum() == TxnType.AssetTransfer,
+                asset_balance.value() == Gtxn[1].asset_amount()
+            ),
+        ),
+        # Increment `BondApp.epoch`
+        App.globalPut(epoch, Add(App.globalGet(epoch), Int(1))),
+        # set `BondApp.current_bond = B_{i+1}`.
+        App.globalPut(current_bond, Gtxn[0].xfer_asset()),
+        Return(Int(1))
+    ])
+
+    program = Cond(
+        # Verfies that the application_id is 0, jumps to on_initialize.
+        [Txn.application_id() == Int(0), on_initialize],
+        # Verifies Update transaction, rejects it.
+        [Txn.on_completion() == OnComplete.UpdateApplication, Return(Int(0))],
+        # Verifies delete transaction, rejects it.
+        [Txn.on_completion() == OnComplete.DeleteApplication, Return(Int(0))],
+        # Verifies closeout transaction, approves it.
+        [Txn.on_completion() == OnComplete.CloseOut, Return(Int(1))],
+        # Verifies opt-in transaction, approves it.
+        [Txn.on_completion() == OnComplete.OptIn, Return(Int(1))],
```

Comment:
only manager should be able to opt-in the bond app to some other asset (otherwise someone could spam and block the bond app).

Reply:
What do you mean by `opt-in the bond app to some other asset` ?
in this condition user can opt-in to the app, this is used when user buys bond tokens. user cannot modify any global variables.

Reply:
This condition is used to opt-int `BondApp` to `B_1`, `B_2`, ... ASA, right? If so, only manager should be able to do it.

Reply:
no, this condition if for an address to opt-in to an app, for tracking B_1, B_2,.. we have a global variable `current_bond`.

Reply:
What would be a condition for the BondApp to opt-in to an ASA?

Reply:
BondApp itself cannot opt-in or send, receive assets/algo's becuase it doesn't have address, so we have issuer, dex lsig which opts-in to assets or send or receive them, In lsig we have check for calling specific app ID, which checks for the global variables/owner, 
- got this case, will add one more condition in lsigs `a algo payment with zero amount from storeManager` then only allow opt-in

Reply:
Done. added this check in all lsigs
```
# allow opt-in transaction
    opt_in = And(
        Gtxn[0].type_enum() == TxnType.Payment,
        Gtxn[0].amount() == Int(0),
        Gtxn[0].sender() == Tmpl.Addr("TMPL_STORE_MANAGER"),
        Gtxn[1].type_enum() == TxnType.AssetTransfer,
        Gtxn[1].asset_amount() == Int(0)
    )
```

Reply:
this is not done - `opt_in` condition should be added here.

Reply:
Ah, my mistake again. App can't `opt-in` to another thing.